### PR TITLE
ContextNode: groups API

### PR DESCRIPTION
**Pull request by dvoytenko**
This API allows to group children of the same node into a sub-node. This makes it easier to to set properties on a whole group rather than doing it 1-by-1. For instance, this allows us to express something like this: "all unslotted children of a node have to be paused".

### CODE REVIEW

**Comment by samouri** [comment]
are there other use cases other than unslotted children? whats the motivation for weight?

**Comment by dvoytenko** [comment]
@samouri yes, there were couple more: placeholder and fallback groups, and possibly loading indicator group. Some of them are present in https://github.com/ampproject/amphtml/pull/29904, which is where I ran into issues of possible overlap. It's generally possible to avoid it with careful selectors. But then I felt that the `weight` is a lot simpler. These groups were "something specific A", "something specific B", and then "everything else". So, the main reason for `weight` is to be able to specify `-1` for unslotted children.

File: src/context/node.js [hunk 1]
```diff
@@ -318,6 +339,55 @@ export class ContextNode {
...
+    const group = groups && groups.get(name);
+    return (group && group.cn) || null;
+  }
+
+  /**
+   * @param {!Node} node
+   * @return {?ContextNode}
+   * @protected
+   */
+  matchGroup(node) {
```
Comment (samouri) [inline src/context/node.js:375]: nit: maybe a name like `computeActiveGroup`? I'm used to a `match` func returning bool

**Review by samouri — APPROVED** [review]

**Comment by samouri** [comment]
Discussed offline regarding some interesting restrictions in present API:
- `group` works by creating an extra intermediary ContextNode and assigning all matching children to have this new CN as its parent. The intermediary has the original CN as its parent so values should still flow as expected. You can then set props/remove them on the group since at that point its just a normal CN.
- A child may only belong to a single group
- When there are multiple groups possible for a child, the highest weight wins (similar to css specificity)

File: src/context/node.js (hunk 1 above)
Comment (dvoytenko) [reply src/context/node.js]: Renamed to `findGroup`

**Pull request merged**